File: recursion.js
function myRange(start, finish){
  if (start === finish){
    return [start];
  }

  return [start].concat(myRange(start + 1, finish));

}
// console.log(myRange(1,5));

Array.prototype.mySum = function() {
  if (this.length === 1){
    return this[0];
  }
    return this[0] + this.slice(1).mySum();
};

// console.log([1,2,3,4,5].mySum());



function myExp(base, pwr) {
  if (pwr === 0) {
    return 1;
  }

  return base * myExp(base, pwr - 1);
}

// console.log(myExp(3, 3));

function fibonacci(n) {
  if (n < 2) {
    return [n];
  }
  if (n === 2) {
    return [1, 1];
  }

  let prev = fibonacci(n - 1);
  let next = prev[prev.length - 2] + prev[prev.length - 1];
  return prev.concat([next]);
}

// console.log(fibonacci(6));

Array.prototype.bSearch = function(target) {
  if (this.length < 1) {
    return null;
  }

  let middle = Math.floor(this.length / 2);

  if (this[middle] === target) {
    return middle;
  }
  else if (target < this[middle]) {
    return this.slice(0, middle).bSearch(target);
  }
  else {
    let res = this.slice(middle + 1, this.length).bSearch(target);
    return (res !== null ? middle + 1 + res : null);
  }

};

// console.log([1, 2, 3].bSearch(1));
// console.log([2, 3, 4, 5].bSearch(3));
// console.log([2, 4, 6, 8, 10].bSearch(6));
// console.log([1, 3, 4, 5, 9].bSearch(5));
// console.log([1, 2, 3, 4, 5, 6].bSearch(6));
// console.log([1, 2, 3, 4, 5, 6].bSearch(0));
// console.log([1, 2, 3, 4, 5, 7].bSearch(6));

function makeChange(amt, coins) {
  if (amt === 0 ) {
    return [];
  }
  if (coins.length === 0) {
    return null;
  }
  // return amout if the amount is the same as a coin value
  if (coins.includes(amt)) {
    return [amt];
  }
  // best coin is the biggest coin that is less than the amount
  let bestCoin = 0;
  coins.forEach((coin) => {
    if (coin > bestCoin && coin < amt) {
      bestCoin = coin;
    }
  });
  // if none of the coins are less than the amt return null
  if (bestCoin === 0) {
    return null;
  }
  let remaining = amt - bestCoin;
  return [bestCoin].concat(makeChange(remaining, coins));
}

// console.log(makeChange(14, [10, 7, 1]));


Array.prototype.mySelect = function(cb) {
  let selected = [];

  this.forEach((el) => {
    if (cb(el) === true) {
      selected.push(el);
    }
  });

  return selected;
};

// console.log([1,2,3,4,5].mySelect((el) => el % 2 === 0));

Array.prototype.bubbleSort = function(cb) {
  let sorted = false;
  while (sorted === false) {
    sorted = true;
    for (let i = 0; i < this.length - 1; i++) {
      if (cb(this[i], this[i + 1]) === true) {
        let temp = this[i];
        this[i] = this[i + 1];
        this[i + 1] = temp;
        sorted = false;
      }
    }
  }
  return this;
};


function makeBetterChange(amt, coins) {
  if (amt === 0 ) {
    return [];
  }
  if (coins.length === 0) {
    return null;
  }
  if (coins.includes(amt)) {
    return [amt];
  }

  let bestAnswer = null;
  let validCoins = coins.mySelect((coin) => coin < amt);

  // go through each coin
   for (let i = 0; i < validCoins.length; i++) {
    let coin = validCoins[i];

    // othercoins = find the make better change for the remaining amount
    let otherCoins = makeBetterChange((amt - coin), coins);
    // next if othercoins is null
    if (otherCoins === null) {
      continue;
    }

    // potential answer is current coin + othercoins
    let potentialAnswer = [coin].concat(otherCoins);

    // save it to best answer if it is smaller in num coins than best answer
    if (bestAnswer === null || bestAnswer.length > potentialAnswer.length) {
      bestAnswer = potentialAnswer;
    }
  }

  // return best answer
  return bestAnswer;
}

// console.log(makeBetterChange(14, [10, 7, 1]));





Array.prototype.mergeSort = function(cb) {
if (this.length === 1) {
  return this;
}

let middle = Math.floor(this.length / 2);
let left = this.slice(0, middle).mergeSort(cb);
let right = this.slice(middle, this.length).mergeSort(cb);

// the sorted halves
  return merge(left, right, cb);
};

function merge(arr1, arr2, cb) {
  let merged = [];
  while(arr1.length > 0 && arr2.length > 0) {
    switch (cb(arr1[0], arr2[0])){
      case -1:
        merged.push(arr1.shift());
        break;
      case 1:
        merged.push(arr2.shift());
        break;
      default:
        merged.push(arr1.shift());
        break;
    }
  }
  let both = arr1.concat(arr2);
  return merged.concat(both);
}

console.log([3,4,7,88,42,122,56,9,63].mergeSort((el1, el2) => {
  if (el1 < el2) {
    return -1;
  }
 else if (el1 > el2){
   return 1;
 }
 else {
   return 0;
 }
}));














//

File: w1d3/recursion.rb
require 'byebug'

def range(min, max)
  return [] if max < min
  [min] + range(min + 1, max)
end

def range_iter(min, max)
  return [] if max < min
  n = min
  final_range = []
  until n > max
    final_range << n
    n += 1
  end
  final_range
end

def sum_array_iter(nums)
  sum = 0
  nums.each { |el| sum += el }
  sum
end

def sum_array(nums)
  return 0 if nums.empty?
  nums.first + sum_array(nums[1..-1])
end

def exponent_1(num, exp)
  return 1 if exp == 0
  num * exponent_1(num, exp - 1)
end

# def exponent_2(num, exp)
#   return 1 if exp == 0
#   return num if exp == 1
#   if exp.even?
#     new_num = exponent_2(num, exp / 2)
#     new_num * new_num
#   else
#     new_num = exponent_2(num, (exp - 1) / 2)
#     num * new_num * new_num
#   end
# end

#SOLUTION:
def exp2(base, power)
  return 1 if power == 0

  half = exp2(base, power / 2)

  if power.even?
    half * half
  else
    # note that (power / 2) == ((power - 1) / 2) if power.odd?
    base * half * half
  end
end

def deep_dup(arr)
  duplicate = []
  arr.each do |el|
    if el.is_a?(Array)
      duplicate << deep_dup(el)
    else
      duplicate << el
    end
  end
  duplicate
end

#ONE LINE SOLUTIONS
def dd_inject
  inject([]) { |dup, el| dup << (el.is_a?(Array) ? dd_inject(el) : el) }
end

# Beware map! The ultimate one-liner.
def dd_map
  map { |el| el.is_a?(Array) ? dd_map(el) : el }
end

def fibonacci_iter(n)
  return [] if n == 0
  return [0] if n == 1
  fibs = [0, 1]

  until fibs.length > n
    fibs << (fibs[-2] + fibs[-1])
  end
  fibs
end

def fibonacci(n)
  return [0, 1].take(n) if n <= 2
  previous = fibonacci(n - 1)
  previous << (previous[-2] + previous[-1])
end

def permutations(arr)
  return [arr] if arr.length <= 1

  perms = permutations(arr[0..-2])
  digit = arr[-1]
  all_perms = []

  (0...arr.length).each do |idx|
    perms.each do |subarray|
      all_perms << subarray[0...idx] + [digit] + subarray[idx..-1]
      # all_perms << subarray.dup.insert(idx, digit)
    end
  end
  all_perms
end

#bsearch solution which is totally cool
# def bsearch(nums, target)
#   # nil if not found; can't find anything in an empty array
#   return nil if nums.empty?
#
#   probe_index = nums.length / 2
#   case target <=> nums[probe_index]
#   when -1
#     # search in left
#     bsearch(nums.take(probe_index), target)
#   when 0
#     probe_index # found it!
#   when 1
#     # search in the right; don't forget that the right subarray starts
#     # at `probe_index + 1`, so we need to offset by that amount.
#     sub_answer = bsearch(nums.drop(probe_index + 1), target)
#     (sub_answer.nil?) ? nil : (probe_index + 1) + sub_answer
#   end
#
#   # Note that the array size is always decreasing through each
#   # recursive call, so we'll either find the item, or eventually end
#   # up with an empty array.
# end


def bsearch(array, target)
  return nil if array.length <= 1 && array[0] != target

  mid_idx = (array.length - 1) / 2
  if array[mid_idx] == target
    mid_idx
  elsif array[mid_idx] < target
    response = bsearch(array[(mid_idx + 1)..-1], target)
    response.nil? ? nil : response + mid_idx + 1
  else
    bsearch(array[0...mid_idx], target)
  end
end

def merge_sort(arr)
  return arr if arr.length <= 1
  mid = (arr.length - 1) / 2

  arr1 = merge_sort(arr[0..mid])
  arr2 = merge_sort(arr[(mid + 1)..-1])

  merge(arr1, arr2)
end

#SOLUTION
def merge(left, right)
  merged_array = []
  until left.empty? || right.empty?
    merged_array <<
      ((left.first < right.first) ? left.shift : right.shift)
  end

  merged_array + left + right
end

# def merge(arr1, arr2)
#   merged = []
#   (arr1.length + arr2.length).times do
#     if arr1.empty?
#       return merged + arr2
#     elsif arr2.empty?
#       return merged + arr1
#     elsif arr1[0] > arr2[0]
#       merged << arr2.shift
#     else
#       merged << arr1.shift
#     end
#   end
#   merged
# end

def subsets(arr)
  return [[]] if arr.empty?
  subsets_arr = []
  cur_el = arr[0]
  prior_subs = subsets(arr[1..-1])
  prior_subs.each do |subset|
    subsets_arr << subset.dup.push(cur_el)
  end
  subsets_arr + prior_subs
end

#SOLUTION
# def subsets(arr)
#   return [[]] if arr.empty?
#   subs = subsets(arr.take(arr.count - 1))
#   subs.concat(subs.map { |sub| sub + [arr.last] })
# end

def greedy_make_change(amt, coins)
  return [] if target == 0
  return nil if coins.none? { |coin| coin <= target }
  return [amt] if coins.include?(amt)
  best_coin = coins.reject { |coin| coin > amt }.max
  remaining = amt - best_coin
  [best_coin] + greedy_make_change(remaining, coins)
end

def make_better_change(amt, coins)
  return [amt] if coins.include?(amt)
  return [] if amt == 0
  best_answer = nil
  valid_coins = coins.reject { |coin| coin > amt}
  valid_coins.each do |coin|
    other_coins = make_better_change((amt - coin), coins)
    next if other_coins.nil?
    potential_answer = [coin] + other_coins
    best_answer = potential_answer if best_answer.nil? || potential_answer.length < best_answer.length
  end
  best_answer
end
